### PR TITLE
Add environment variable to set prefered UI-backend

### PR DESCRIPTION
**Pull request by besser82**
Add `$YUI_PREFERED_BACKEND` to set prefered UI-backend, because some window-managers don't indicate `$XDG_CURRENT_DESKTOP`. This way users can set their wanted backend in `~/.bash_profile`, instead of needing to deal with cli-switches.

### CODE REVIEW

**Comment by opensusegithubapi** [comment]
Can one of the admins verify this patch?

File: src/YUILoader.cc [hunk 1]
```diff
@@ -58,74 +76,88 @@ void YUILoader::loadUI( bool withThreads )
...
-	yuiMilestone () << "Using UI-backend: \"" << YUIPlugin_Qt << "\". Forced on command-line." << std::endl;
-    bool wantGtk = cmdline.find("--gtk") != -1;
-    if( wantGtk )
-	yuiMilestone () << "Using UI-backend: \"" << YUIPlugin_Gtk << "\". Forced on command-line." << std::endl;
+    bool wantGtk     = ( cmdline.find( "--gtk" )     != -1 );
+    bool wantNcurses = ( cmdline.find( "--ncurses" ) != -1 );
+    bool wantQt      = ( cmdline.find( "--qt" )      != -1 );
+    wantGtk     = ( wantGtk     || ( strcmp( envPreset, YUIPlugin_Gtk )     == 0 ) ) && !( wantNcurses || wantQt );
+    wantNcurses = ( wantNcurses || ( strcmp( envPreset, YUIPlugin_NCurses ) == 0 ) ) && !( wantGtk     || wantQt );
+    wantQt      = ( wantQt      || ( strcmp( envPreset, YUIPlugin_Qt )      == 0 ) ) && !( wantGtk     || wantNcurses );
```
Comment (jreidinger) [inline src/YUILoader.cc]: these conditions are quite hell, what about
```
bool wantGtk     = ( cmdline.find( "--gtk" )     != -1 );
bool wantNcurses = ( cmdline.find( "--ncurses" ) != -1 );
￼bool wantQt      = ( cmdline.find( "--qt" )      != -1 );
boot wanted_ui = wantGtk || wantNcurses || wantQt;
if (!wanted_ui)
{
  wantGtk     =  strcmp( envPreset, YUIPlugin_Gtk )     == 0;
  ...
}
```

Comment (besser82) [inline src/YUILoader.cc]: Yes, didn't see the easier way, when I coded this…  I fixed it in the rebased commits.

**Comment by besser82** [comment]
The changes from the review should be all in the recent commits.

File: src/YUILoader.cc [hunk 2]
```diff
@@ -42,11 +42,29 @@
...
+
+    if( getenv( "XDG_CURRENT_DESKTOP" ) )
+	envDesktop = getenv( "XDG_CURRENT_DESKTOP" );
+    else
+	envDesktop = "";
+
+    if( getenv( "YUI_PREFERED_BACKEND" ) )
+	envPreset = getenv( "YUI_PREFERED_BACKEND" );
+    else
+	envPreset = "";
```
Comment (jreidinger) [inline src/YUILoader.cc]: one more idea, you can simplify this code with
```
const char *envPreset = getenv( "YUI_PREFERED_BACKEND" ) || "";
```

Comment (jreidinger) [inline src/YUILoader.cc]: and similar for above code.

File: src/YUILoader.cc [hunk 3]
```diff
@@ -58,74 +76,93 @@ void YUILoader::loadUI( bool withThreads )
...
-    bool wantGtk = cmdline.find("--gtk") != -1;
-    if( wantGtk )
-	yuiMilestone () << "Using UI-backend: \"" << YUIPlugin_Gtk << "\". Forced on command-line." << std::endl;
+    bool haveGtk     = pluginExists( YUIPlugin_Gtk );
+    bool haveNcurses = pluginExists( YUIPlugin_NCurses );
+    bool haveQt      = pluginExists( YUIPlugin_Qt );
 
-    bool haveQt = pluginExists( YUIPlugin_Qt );
-    bool haveGtk = pluginExists( YUIPlugin_Gtk );
+    wantedGUI="";
```
Comment (jreidinger) [inline src/YUILoader.cc]: this look like bug in a logic. on line 97-99 you set wantedGUI and here you assign it again to empty.

**Comment by jreidinger** [comment]
@besser82 thanks, I found one more issue with code and one idea how to simplify it.

File: src/YUILoader.cc (hunk 3 above)
Comment (besser82) [inline src/YUILoader.cc]: No, it's not an error in logic.  In line 97-99 it is set to get the possible user-selected UI for YMilestone() in line 101.

I reset it here, so the loader can work it's magic on picking an UI-plugin as described in the docs.

File: src/YUILoader.cc (hunk 2 above)
Comment (besser82) [inline src/YUILoader.cc]: Fixed that in upcoming rebased commits.

**Comment by besser82** [comment]
Applied your idea for the first finding.  Your second finding is fine the way it is; I commented about it in your review.

**Comment by besser82** [comment]
I changed the initialization to a one-liner; your proposal led to an error:
```
src/YUILoader.cc:47:67: error: cannot convert 'bool' to 'const char*' in initialization
     const char * envPreset  = getenv( "YUI_PREFERED_BACKEND" ) || "";
```

**Comment by jreidinger** [comment]
@besser82 hmmm, typo check, so you can use ` const char * envPreset  = getenv( "YUI_PREFERED_BACKEND" ) ?: "";`

**Comment by jreidinger** [comment]
For second finding, I find it a bit confusing, can you at least comment it in code, that reset is intentional?

**Comment by besser82** [comment]
Your changed works now and I added the comment in rebased commits.

**Comment by besser82** [comment]
@jreidinger Thank you for the review and your suggestions!  =)

**Pull request merged**